resolver.go

```diff
@@ -308,12 +308,97 @@ func setFields(ctx *hcontext, req *http.Request, input reflect.Value, t reflect.
 			inField.Set(reflect.ValueOf(parsed))
 		}
 	}
+}
+
+// A smart join for JSONPath
+func pathJoin(prefix string, parts ...string) string {
+	joined := prefix
+	if joined != "" {
+		joined += "."
+	}
+	return joined + strings.Join(parts, ".")
+}
+
+// ctxLocationWrapper wraps a context so that the error detail `location` field
+// gets sets appropriately for resolver errors. I.e. the resolver doesn't know
+// when it runs whether it is the body or deeply nested within the body of an
+// incoming request. We prefix it so the errors make sense to the end-user.
+type ctxLocationWrapper struct {
+	*hcontext
+	location string
+}
+
+func (c ctxLocationWrapper) AddError(err error) {
+	if e, ok := err.(*ErrorDetail); ok {
+		e.Location = pathJoin(c.location, e.Location)
+	}
+
+	c.hcontext.AddError(err)
+}
+
+// resolveFields recursively crawls the input struct and calls Resolve on
+// any structs it finds as fields, within slices, and as values in maps. This
+// should be called *after* all other fields are set so the resolver code can
+// use their values. It processes depth-first so structs have access to the
+// resolved fields of any contained structs when their resolver runs.
+func resolveFields(ctx *hcontext, path string, input reflect.Value) {
+	if input.Kind() == reflect.Ptr {
+		input = input.Elem()
+	}
+	if input.Kind() == reflect.Invalid {
+		// Some internal stuff can return invalid, e.g. time.Time fields. We just
+		// ignore those.
+		return
+	}
 
-	// Resolve after all other fields are set so the resolver can use them,
-	// and also so that any embedded structs are resolved first.
-	if input.CanInterface() {
+	// First, handle any nested stuff (depth-first search)
+	switch input.Kind() {
+	case reflect.Slice:
+		for i := 0; i < input.Len(); i++ {
+			resolveFields(ctx, fmt.Sprintf("%s[%d]", path, i), input.Index(i))
+		}
+	case reflect.Map:
+		keys := input.MapKeys()
+		for i := 0; i < input.Len(); i++ {
+			resolveFields(ctx, pathJoin(path, keys[i].String()), input.MapIndex(keys[i]))
+		}
+	case reflect.Struct:
+		for i := 0; i < input.NumField(); i++ {
+			f := input.Type().Field(i)
+			n := strings.ToLower(f.Name)
+
+			if j, ok := f.Tag.Lookup("json"); ok {
+				parts := strings.Split(j, ",")
+				if parts[0] != "" {
+					n = parts[0]
+				}
+			}
+
+			if path == "" {
+				// Check what kind of top-level path there should be, if any. This
+				// will get errors where the location is e.g. query.search or
+				// header.authorization so you know where to look.
+				for _, tag := range []string{"path", "query", "header"} {
```

Comment:
This looks like a list of magic strings; maybe declare them above with comments linking back to docs explaining them?